R/write_output.R

```diff
@@ -0,0 +1,338 @@
+#' @param output_dir String. The base output directory path.
+#' @param samples A data.table as returned by [process_samples()]
+#' @param summaries A data.table as returned by [process_quantiles()]
+#' @param job_id String. The identifier for the job.
+#' @param task_id String. The identifier for the task.
+#' @param metadata List. Additional metadata to be included in the output.
+#'
+#' @return Invisible NULL. The function is called for its side effects.
+#' @export
+write_model_outputs <- function(
+    fit,
+    samples,
+    summaries,
+    output_dir,
+    job_id,
+    task_id,
+    metadata = list()) {
+  rlang::try_fetch(
+    {
+      # Create directory structure
+      write_output_dir_structure(output_dir, job_id, task_id)
+
+      # Write raw samples
+      samples_path <- file.path(
+        output_dir,
+        job_id,
+        "samples",
+        paste0(task_id, ".parquet")
+      )
+      write_parquet(samples, samples_path)
+      cli::cli_alert_success("Wrote samples to {.path {samples_path}}")
+
+      # Process and write summarized quantiles
+      summaries_path <- file.path(
+        output_dir,
+        job_id,
+        "summaries",
+        paste0(task_id, ".parquet")
+      )
+      write_parquet(summaries, summaries_path)
+      cli::cli_alert_success("Wrote summaries to {.path {summaries_path}}")
+
+
+      # Write EpiNow2 model
+      model_path <- file.path(
+        output_dir,
+        job_id,
+        "tasks",
+        task_id,
+        "model.rds"
+      )
+      saveRDS(fit, model_path)
+      cli::cli_alert_success("Wrote model to {.path {model_path}}")
+
+
+      # Write model run metadata
+      metadata_path <- file.path(
+        output_dir,
+        job_id,
+        "tasks",
+        task_id,
+        "metadata.json"
+      )
+      jsonlite::write_json(metadata, metadata_path, pretty = TRUE)
+      cli::cli_alert_success("Wrote metadata to {.path {metadata_path}}")
+    },
+    error = function(cnd) {
+      # Downgrade erroring out to a warning so we can catch and log
+      cli::cli_warn(
+        "Failure writing outputs",
+        parent = cnd,
+        class = "no_outputs"
+      )
+    }
+  )
+
+  invisible(NULL)
+}
+
+#' Create output directory structure for a given job and task.
+#'
+#' This function generates the necessary directory structure for storing output
+#' files related to a job and its tasks, including directories for raw samples
+#' and summarized quantiles.
+#'
+#' @param output_dir String. The base output directory path.
+#' @param job_id String. The identifier for the job.
+#' @param task_id String. The identifier for the task.
+#'
+#' @return The path to the base output directory (invisible).
+#' @export
+write_output_dir_structure <- function(output_dir, job_id, task_id) {
+  # Top level
+  dir.create(output_dir,
+    showWarnings = FALSE
+  )
+  # Second level
+  dir.create(file.path(output_dir, job_id),
+    showWarnings = FALSE
+  )
+  # Third level
```

Comment:
Is there a way to do all of tasks, samples and summaries in a shorter call?

Reply:
https://github.com/CDCgov/cfa-epinow2-pipeline/pull/41/commits/c80c8cff89378607fdf464260b00f6d5e918faae